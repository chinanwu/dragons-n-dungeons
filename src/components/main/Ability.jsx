import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { ABILITY_SKILLS } from '../../constants/Abilities';
import { applyAbilityProxy } from '../../thunk/AbilityThunk.jsx';
import { defaultScore } from '../../reducers/AbilityReducer';
import Input from '../general/Input.jsx';
import { Skill } from './Skill.jsx';

import './Ability.less';

export const Ability = ({ id, name, score, modifier, onChange }) => {
  const handleChange = useCallback(
    event => {
      if (event && event.target) {
        if (event.target.value) {
          let num = parseInt(event.target.value);
          if (num) {
            // TODO: If greater than max or less than min, do special effect?
            num = num > 30 ? 30 : num < 0 ? 0 : num;
            onChange(name, num);
          }
        } else {
          onChange(name, defaultScore);
        }
      }
    },
    [onChange]
  );

  return (
    <div id={id} className="Ability">
      <div className="AbilityName">{name}</div>
      <div className="AbilityContent">
        <div className="Ability--left">
          <Input
            id="abilityScore"
            className="AbilityScore"
            type="text"
            maxLength={2}
            value={score}
            onChange={handleChange}
          />
          <div className="AbilityModifier">{modifier}</div>
        </div>
        <div className="Ability--right">
          {ABILITY_SKILLS[name.toLowerCase()].map(skill => (
            <Skill
              id={name.toLowerCase() + 'Skill'}
              name={skill}
              modifier={modifier}
              key={skill}
            />
          ))}
        </div>
      </div>
    </div>
  );
};

Ability.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string,
  score: PropTypes.number,
  modifier: PropTypes.number,
  onChange: PropTypes.func,
};

// TODO: Maybe in the future, instead of having to make everything lowercase,
//  it is set as lowercase in the constants file, and css styling will deal with it?
export const mapStateToProps = ({ ability }, { name }) => ({
  score: ability[name.toLowerCase()].score,
  modifier: ability[name.toLowerCase()].modifier,
});

export const mapDispatchToProps = {
  onChange: applyAbilityProxy,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Ability);
