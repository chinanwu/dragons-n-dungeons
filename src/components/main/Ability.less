@import '../styling/spacing.less';

.Ability {
  width: 230px;
  height: 170px;
  border: 1px solid black;
}
.AbilityName {
  text-align: center;
  margin-top: 6px;
}

.AbilityContent {
  display: flex;
  margin-top: 8px;
}

.Ability--left {
  display: block;
  width: 30%;
  text-align: center;
}

.AbilityScore {
  background-color: inherit;
  font-size: @larger-font;
  border: none;
  width: inherit;
}

input::-webkit-outer-spin-button,
input::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

input[type='number'] {
  -moz-appearance: textfield;
}

.AbilityModifier {
  font-size: 21px;
}

.Ability--right {
  display: flex;
  flex-direction: column;
}
