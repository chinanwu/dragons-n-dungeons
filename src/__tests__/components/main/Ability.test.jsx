import { shallow } from 'enzyme';
import React from 'react';

import { Ability, mapStateToProps } from '../../../components/main/Ability.jsx';

jest.unmock('../../../components/main/Ability.jsx');
jest.unmock('../../../constants/Abilities');

describe('Ability component', () => {
  describe('rendering', () => {
    it('renders Ability', () => {
      const wrapper = shallow(<Ability id="test" name="Strength" />);
      expect(wrapper.getElement()).toMatchSnapshot();
    });
  });

  describe('behaviour', () => {
    it('saves changed input', () => {
      const onChange = jest.fn();
      const event = {
        target: {
          value: '1',
        },
      };
      const wrapper = shallow(
        <Ability id="test" name="Strength" onChange={onChange} />
      );
      expect(wrapper.getElement()).toMatchSnapshot();
      wrapper.find('#abilityScore').simulate('change', event);
      expect(onChange.mock.calls).toEqual([['Strength', 1]]);
    });

    it('changes out of bounds input to max or min', () => {
      const onChange = jest.fn();
      const smallEvent = {
        target: {
          value: '-1',
        },
      };
      const largeEvent = {
        target: {
          value: '31',
        },
      };
      const wrapper = shallow(
        <Ability id="test" name="Strength" onChange={onChange} />
      );
      expect(wrapper.getElement()).toMatchSnapshot();
      wrapper.find('#abilityScore').simulate('change', smallEvent);
      wrapper.find('#abilityScore').simulate('change', largeEvent);
      expect(onChange.mock.calls).toEqual([['Strength', 0], ['Strength', 30]]);
    });

    it('changes null input to default score', () => {
      const onChange = jest.fn();
      const event = {
        target: {
          value: null,
        },
      };
      const wrapper = shallow(
        <Ability id="test" name="Strength" onChange={onChange} />
      );
      expect(wrapper.getElement()).toMatchSnapshot();
      wrapper.find('#abilityScore').simulate('change', event);
      expect(onChange.mock.calls).toEqual([['Strength', 0]]);
    });

    it('does nothing  if invalid input', () => {
      const onChange = jest.fn();
      const event = {
        target: {
          value: 'a',
        },
      };
      const wrapper = shallow(
        <Ability id="test" name="Strength" onChange={onChange} />
      );
      expect(wrapper.getElement()).toMatchSnapshot();
      wrapper.find('#abilityScore').simulate('change', event);
      expect(onChange).not.toHaveBeenCalled();
    });

    it('does nothing if event is null', () => {
      const onChange = jest.fn();
      const wrapper = shallow(
        <Ability id="test" name="Strength" onChange={onChange} />
      );
      expect(wrapper.getElement()).toMatchSnapshot();
      wrapper.find('#abilityScore').simulate('change', null);
      expect(onChange).not.toHaveBeenCalled();
    });
  });

  describe('mapStateToProps', () => {
    it('returned mapped properties', () => {
      const state = { test: { score: 1, modifier: 2 } };
      expect(mapStateToProps({ ability: state }, { name: 'Test' })).toEqual({
        score: 1,
        modifier: 2,
      });
    });
  });
});
