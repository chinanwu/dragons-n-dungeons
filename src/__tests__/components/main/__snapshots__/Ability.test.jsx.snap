// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ability component behaviour changes null input to default score 1`] = `
<div
  className="Ability"
  id="test"
>
  <div
    className="AbilityName"
  >
    Strength
  </div>
  <div
    className="AbilityContent"
  >
    <div
      className="Ability--left"
    >
      <input
        className="AbilityScore"
        id="abilityScore"
        maxLength={2}
        onChange={[Function]}
        type="number"
      />
      <div
        className="AbilityModifier"
      />
    </div>
    <div
      className="Ability--right"
    >
      <Skill
        id="strengthSkill"
        name="Saving Throws"
      />
      <Skill
        id="strengthSkill"
        name="Athletics"
      />
    </div>
  </div>
</div>
`;

exports[`Ability component behaviour changes out of bounds input to max or min 1`] = `
<div
  className="Ability"
  id="test"
>
  <div
    className="AbilityName"
  >
    Strength
  </div>
  <div
    className="AbilityContent"
  >
    <div
      className="Ability--left"
    >
      <input
        className="AbilityScore"
        id="abilityScore"
        maxLength={2}
        onChange={[Function]}
        type="number"
      />
      <div
        className="AbilityModifier"
      />
    </div>
    <div
      className="Ability--right"
    >
      <Skill
        id="strengthSkill"
        name="Saving Throws"
      />
      <Skill
        id="strengthSkill"
        name="Athletics"
      />
    </div>
  </div>
</div>
`;

exports[`Ability component behaviour does nothing  if invalid input 1`] = `
<div
  className="Ability"
  id="test"
>
  <div
    className="AbilityName"
  >
    Strength
  </div>
  <div
    className="AbilityContent"
  >
    <div
      className="Ability--left"
    >
      <input
        className="AbilityScore"
        id="abilityScore"
        maxLength={2}
        onChange={[Function]}
        type="number"
      />
      <div
        className="AbilityModifier"
      />
    </div>
    <div
      className="Ability--right"
    >
      <Skill
        id="strengthSkill"
        name="Saving Throws"
      />
      <Skill
        id="strengthSkill"
        name="Athletics"
      />
    </div>
  </div>
</div>
`;

exports[`Ability component behaviour does nothing if event is null 1`] = `
<div
  className="Ability"
  id="test"
>
  <div
    className="AbilityName"
  >
    Strength
  </div>
  <div
    className="AbilityContent"
  >
    <div
      className="Ability--left"
    >
      <input
        className="AbilityScore"
        id="abilityScore"
        maxLength={2}
        onChange={[Function]}
        type="number"
      />
      <div
        className="AbilityModifier"
      />
    </div>
    <div
      className="Ability--right"
    >
      <Skill
        id="strengthSkill"
        name="Saving Throws"
      />
      <Skill
        id="strengthSkill"
        name="Athletics"
      />
    </div>
  </div>
</div>
`;

exports[`Ability component behaviour saves changed input 1`] = `
<div
  className="Ability"
  id="test"
>
  <div
    className="AbilityName"
  >
    Strength
  </div>
  <div
    className="AbilityContent"
  >
    <div
      className="Ability--left"
    >
      <input
        className="AbilityScore"
        id="abilityScore"
        maxLength={2}
        onChange={[Function]}
        type="number"
      />
      <div
        className="AbilityModifier"
      />
    </div>
    <div
      className="Ability--right"
    >
      <Skill
        id="strengthSkill"
        name="Saving Throws"
      />
      <Skill
        id="strengthSkill"
        name="Athletics"
      />
    </div>
  </div>
</div>
`;

exports[`Ability component rendering renders Ability 1`] = `
<div
  className="Ability"
  id="test"
>
  <div
    className="AbilityName"
  >
    Strength
  </div>
  <div
    className="AbilityContent"
  >
    <div
      className="Ability--left"
    >
      <input
        className="AbilityScore"
        id="abilityScore"
        maxLength={2}
        onChange={[Function]}
        type="number"
      />
      <div
        className="AbilityModifier"
      />
    </div>
    <div
      className="Ability--right"
    >
      <Skill
        id="strengthSkill"
        name="Saving Throws"
      />
      <Skill
        id="strengthSkill"
        name="Athletics"
      />
    </div>
  </div>
</div>
`;
